firmware/src/rdm_responder.c

```diff
@@ -0,0 +1,427 @@
+    for (j = 0; j != iov[i].length; j++) {
+      checksum += ((uint8_t*) iov[i].base)[j];
+    }
+  }
+  checksum_data[0] = ShortMSB(checksum);
+  checksum_data[1] = ShortLSB(checksum);
+}
+
+/*
+ * @brief Send a DUB response if our UID falls within the range.
+ */
+static void SendDUBResponseIfRequired(const uint8_t *param_data,
+                                      unsigned int param_data_length) {
+  if (param_data_length != 2 * UID_LENGTH) {
+    return;
+  }
+
+  if (!(UIDCompare(param_data, g_rdm_responder.uid) <= 0 &&
+        UIDCompare(g_rdm_responder.uid, param_data + UID_LENGTH) <= 0)) {
+    return;
+  }
+
+  uint8_t response[DUB_RESPONSE_LENGTH];
+  memset(response, FE_CONSTANT, 7);
+  response[7] = AA_CONSTANT;
+  response[8] = g_rdm_responder.uid[0] | AA_CONSTANT;
+  response[9] = g_rdm_responder.uid[0] | FIVE5_CONSTANT;
+  response[10] = g_rdm_responder.uid[1] | AA_CONSTANT;
+  response[11] = g_rdm_responder.uid[1] | FIVE5_CONSTANT;
+
+  response[12] = g_rdm_responder.uid[2] | AA_CONSTANT;
+  response[13] = g_rdm_responder.uid[2] | FIVE5_CONSTANT;
+  response[14] = g_rdm_responder.uid[3] | AA_CONSTANT;
+  response[15] = g_rdm_responder.uid[3] | FIVE5_CONSTANT;
+  response[16] = g_rdm_responder.uid[4] | AA_CONSTANT;
+  response[17] = g_rdm_responder.uid[4] | FIVE5_CONSTANT;
+  response[18] = g_rdm_responder.uid[5] | AA_CONSTANT;
+  response[19] = g_rdm_responder.uid[5] | FIVE5_CONSTANT;
+
+  uint16_t checksum = 0;
+  unsigned int i;
+  for (i = 8; i < 20; i++) {
+    checksum += response[i];
+  }
+
+  response[20] = ShortMSB(checksum) | AA_CONSTANT;
+  response[21] = ShortMSB(checksum) | FIVE5_CONSTANT;
+  response[22] = ShortLSB(checksum) | AA_CONSTANT;
+  response[23] = ShortLSB(checksum) | FIVE5_CONSTANT;
+
+  IOVec iov;
+  iov.base = response;
+  iov.length = DUB_RESPONSE_LENGTH;
+
+#ifdef PIPELINE_RDMRESPONDER_SEND
+  PIPELINE_RDMRESPONDER_SEND(false, &iov, 1);
+#else
+  g_rdm_responder.send_callback(false, &iov, 1);
+#endif
+}
+
+/*
+ * @brief Send a RDM response, if the request was not broadcast.
+ */
+static void RespondIfRequired(const RDMHeader *incoming_header,
+                              RDMResponseType response_type,
+                              RDMCommandClass command_class,
+                              RDMPid pid,
+                              const uint8_t *param_data,
+                              unsigned int param_data_length) {
+  if (!RequiresResponse(incoming_header)) {
+    return;
+  }
+
+  uint8_t start_code = RDM_START_CODE;
+
+  RDMHeader outgoing_header;
+  outgoing_header.sub_start_code = SUB_START_CODE;
+  outgoing_header.message_length = (
+      sizeof(start_code) + sizeof(outgoing_header) + param_data_length);
+  memcpy(outgoing_header.dest_uid, incoming_header->src_uid, UID_LENGTH);
+  memcpy(outgoing_header.src_uid, g_rdm_responder.uid, UID_LENGTH);
+  outgoing_header.transaction_number = incoming_header->transaction_number;
+  outgoing_header.port_id = response_type;
```

Comment:
It's one field that doubles for both. In the structure it's named after port_id because I don't want to call it 

port_id_or_response_type